WiP: Refactor CharmHelpers

gabriel-samfira opened the pull request:
This is branch refactors CharmHelpers to be more modular. It is still a work in progress and does not retain 100% backwards compatibility. Some functions have been removed.

This branch:
- Adds commandlet bindings to all functions
- adds proper comment based documentation
- adds unit tests
- splits the monolithic CharmHelpers into smaller powershell modules
- fixes broken logic
- adds some new functionallity

Still needs to:
- Create a CharmHelpers module that simply imports all the rest, to be backwards compatible
- add equivalent for the Carbon module that works on Nano server

ader1990 committed:
Synced juju powershell helpers

gabriel-samfira committed:
Merge pull request #8 from ader1990/experimental

Synced juju powershell helpers

gabriel-samfira committed:
add propper logging functions

gabriel-samfira committed:
Refactor some functions

* Added comandlet bindings to some functions.
* Added documentation strings
* added aliases for some parameters
* made some commandlets/parameters Obsolete

WiP

gabriel-samfira committed:
Some more refactoring

gabriel-samfira committed:
finish documenting and converting functions

Converted functions in Juju.psm1 to advanced functions. Once all other modules are aligned, a subsequent commit will be made to augment current functions to allow easier workflow.

gabriel-samfira committed:
Segregate modules

This PR starts segregating the individual modules into smaller modules.

gabriel-samfira committed:
Trim some functions and add bindings

gabriel-samfira committed:
Properly define aliases and exports

gabriel-samfira committed:
Added new modules

* added JujuLoging. This will only contain loging functions. It depends on JujuHelper which will contain functions generally useful to other Juju powershell modules.

* added JujuHelper. Currently it only contains the Invoke-JujuCommand which is used throughout some of the other modules.

* added minimum powershell required version of 3
* added RequiredModules in JujuLoging, JujuHooks and JujuUtils

gabriel-samfira committed:
Added JujuWindowsUtils

Separate windows common utility functions

gabriel-samfira committed:
Rename commandlets and fixes

* Renamed Change-ServiceLogon to Set-ServiceLogon
* made Change-ServiceLogon more robust. It now accepts services as string,ManagementObject and CimInstance. This way you can use it on any version of windows. If you specify the service name as a string, it looks up the service name and calls Change through the proper mechanism.
* Renamed Get-JujuUnitName to Convert-JujuUnitNameToNetbios

gabriel-samfira committed:
Set default username

gabriel-samfira committed:
More refactor

gabriel-samfira committed:
More docstrings, more changes

* Add-WindowsUser now uses net.exe for all operations regarding local user creation. This works on all supported windows platforms.
* Added some helper functions to get user accounts/groups by SID/Name

gabriel-samfira committed:
Various fixes

* Fix module version in psd1 files
* Fix JujuHooks typos

gabriel-samfira committed:
fix JujuUtils typos

gabriel-samfira committed:
modules follows approved verbs

gabriel-samfira committed:
added HTML help

Generated by https://github.com/ChaseFlorell/psDoc

gabriel-samfira committed:
Moved some functions around

gabriel-samfira committed:
Added some tests

gabriel-samfira committed:
added some more tests + fixes

gabriel-samfira committed:
Added more tests

gabriel-samfira committed:
Added badge

gabriel-samfira committed:
added more tests

gabriel-samfira committed:
Added more tests

gabriel-samfira committed:
Added more tests

gabriel-samfira committed:
fixed Start-ExternalCommand

gabriel-samfira committed:
added some more tests

gabriel-samfira committed:
remove extra test cases

gabriel-samfira committed:
update readme

ionutbalutoiu commented on JujuHelper/JujuHelper.psm1:
```diff
@@ -0,0 +1,133 @@
+function Invoke-FastWebRequest {
+    <#
+    .SYNOPSIS
+    Invoke-FastWebRequest downloads a file form the web via HTTP. This function will work on all modern windows versions,
```
Small typo. "form" instead of "from"

ionutbalutoiu commented on JujuHooks/JujuHooks.psm1:
```diff
@@ -0,0 +1,1225 @@
+function Confirm-IP {
+    <#
+    .SYNOPSIS
+     Check if the parameter passed as a string, is a valid IPv4 address
```
It will return `$true` even for valid IPv6 addresses. Need to complete description.

ionutbalutoiu commented on JujuHooks/JujuHooks.psm1:
```diff
@@ -0,0 +1,1225 @@
+    return $dnsServers
+}
+
+function Confirm-JujuPortRangeOpen {
```
I think that the issue with overlapping ranges reported [here](https://github.com/cloudbase/juju-powershell-modules/issues/2) still applies.

gabriel-samfira replied:
It does. In which case the juju tool will error out, and the error gets propagated. No use in catching and throwing without doing anything else.

gabriel-samfira committed:
Added Import-Certificate

gabriel-samfira committed:
moved Write-HookTraceback to JujuLoging

gabriel-samfira committed:
Added Grant-Privilege and fixes

Added a new Grant-Privilege function that uses a pre compiled binary of: https://github.com/cloudbase/SetUserAccountRights
This works on all windows versions including Nano.

Fixed Expand-ZipArchive to allow relative paths

gabriel-samfira committed:
Added powershell-yaml submodule

gabriel-samfira committed:
Replace json with yaml

ConvertTo-Json and ConvertFrom-Json is not yet available on Nano server. We switch to ConvertTo-Yaml and ConvertFrom-Yaml for now.

gabriel-samfira committed:
fix JujuUtils and tests

gabriel-samfira committed:
Add Get-RandomString

gabriel-samfira committed:
update appveyor.yml

gabriel-samfira committed:
Fix typo

gabriel-samfira committed:
fixes

Invoke-JujuCommand has been changed to not error out of a native command writes to stderr instead of stdout.
fixed Add-WindowsUser

gabriel-samfira committed:
some extra checks on values

gabriel-samfira committed:
Added Add-UserToLocalGroup

Added a new commandlet to allow adding a user to a local group. Also changed Get-UserGroupMembership to allow domain user lookup. Accepted formats are:

<NetBIOSName>\<Username>
<username>@<domain>

gabriel-samfira committed:
fix typo

gabriel-samfira committed:
suppress output form cmd

gabriel-samfira committed:
Fix JujuHooks

Powershell tries to be smarter then your average language, and does an automatic Split("`r`n") when assigning the output of a native application to a variable. We need to merge that again before sending it to ConvertFrom-Yaml.

Also added an OptionalContext to Get-JujuRelationContext

gabriel-samfira committed:
remove extra logging

gabriel-samfira committed:
Added Confirm-IsMemberOfGroup commandlet

gabriel-samfira committed:
fix typo

gabriel-samfira committed:
export Get-ManagementObject

gabriel-samfira committed:
Move Get-ManagementObject to JujuHelper

gabriel-samfira committed:
explicitly export functions and aliases

Powershell < 5 does not export aliases and variables by default from modules. As such, even if you do specify the alias in the module manifest under AliasesToExport, they will still not be available.

gabriel-samfira committed:
format exported resources

gabriel-samfira committed:
Added networking module

gabriel-samfira committed:
Added copyright headers

gabriel-samfira committed:
remove ErrorActionPreference from modules

This should be set by application.

gabriel-samfira committed:
Move Test-FileIntegrity to JujuHelper

gabriel-samfira committed:
Remove RequiredModules

Directly import required modules in the root module file. On Windows 2012 R2 powershell throws a cyclic dependency error if you import a module declares 2 required modules, and one of those modules, also requires the second one declared in the imported module.

| m1 -----> m3
 --- m2 ----|

This should not cycle normally. M1 requires M2 and M3. M2 requires M3. When importing M1, powershell should import M2, which in turn imports M3. M1 then imports M3 which should be ignored if already loaded. Nothing in this scenario cycles back. This error does not happen in powershell 5 (present in Windows 10 and Windows server 2016).

gabriel-samfira committed:
respond to feedback

gabriel-samfira committed:
fix unit tests

gabriel-samfira committed:
Merge branch 'experimental' of github.com:cloudbase/juju-powershell-modules into experimental
